監視用traQbotの作成

ras0q opened the pull request:
テストしてないです

ras0q committed:
♻️: 変数名変更

newClient -> NewTraqClient

ras0q committed:
🎉: bot作成

メッセージ作成イベントを受けたらfileIDを探してDBにInsertする

SSlime-s commented on router/bot/handler.go:
```diff
@@ -0,0 +1,65 @@
+		return c.NoContent(http.StatusBadRequest)
+	}
+
+	switch event {
+	case "MESSAGE_CREATED":
+		ctx := c.Request().Context()
+		sess, err := session.Get("sessions", c)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Errorf("Failed In Getting Session:%w", err))
+		}
+		accessToken := sess.Values["accessToken"].(string)
+		payload := &traqbot.MessageCreatedPayload{}
+		err = c.Bind(payload)
+		if err != nil {
+			return c.JSON(http.StatusBadRequest, fmt.Errorf("failed to bind request: %w", err))
+		}
+
+		err = MessageCreatedHandler(ctx, accessToken, payload)
+		if err != nil {
+			return c.JSON(http.StatusInternalServerError, fmt.Errorf("failed to handle event: %w", err))
+		}
```
編集・削除も見たほうが

SSlime-s reviewed (commented):
動作確認してない
ちょっと気になったとこだけ

ras0q committed:
♻️: 共通動作をまとめた

ras0q committed:
✨: DELETEイベントの受け取り

files, favoritesどちらからも削除するようにした

ras0q committed:
🐛: インデックス修正

ras0q committed:
✨: UPDATEDイベントの追加

新しく追加されたファイルは挿入、削除されたファイルはfiles、favoritesテーブルから削除

BREAKING CHANGE:
DELETEDのsql文を一部修正

Ras96 commented:
どっちも追加しました
一応sql文のテストだけはした
削除のときの実装がカスそうなので見てほしい

ras0q committed:
🎨: 追加時に拡張子除く

ras0q committed:
♻️: lintかけた

ras0q committed:
🐛: $つけた

ras0q committed:
♻️: 正規表現で取り出す部分を関数に

SSlime-s commented on model/file.go:
```diff
@@ -196,6 +207,54 @@ func ToggleFileFavorite(ctx context.Context, accessToken string, userID string,
+	return nil
+}
+
+func DeleteFiles(ctx context.Context, messageID string) error {
```
きもいかも
```suggestion
func DeleteFilesFromMessageId(ctx context.Context, messageID string) error {
```

SSlime-s commented on router/bot/message_updated.go:
```diff
@@ -0,0 +1,61 @@
+				CreatedAt:    payload.Message.CreatedAt,
+			}
+
+			err = model.InsertFile(ctx, &req)
+			if err != nil {
+				return fmt.Errorf("failed to insert file: %w", err)
+			}
+		}
+	}
+
+	newMap := map[string]struct{}{}
+	for _, v := range fileIDs {
+		newMap[v] = struct{}{}
+	}
+	oldArr, err := model.GetFileIDsInMessage(ctx, payload.Message.ID)
+	if err != nil {
+		return err
+	}
+	for _, v := range oldArr {
+		if _, ok := newMap[v]; !ok {
+			if err := model.DeleteFile(ctx, v); err != nil {
+				return err
```
まとめて Insert Delete してほしい

SSlime-s commented on router/bot/util.go:
```diff
@@ -0,0 +1,34 @@
+	}
+}
+
+func removeExt(str string) string {
```
略語使わないほうが:yosa:

SSlime-s reviewed (commented):
ちょっと

ras0q committed:
Update model/file.go

Co-authored-by: しゅーまい <62363188+SSlime-s@users.noreply.github.com>

ras0q committed:
♻️: 関数名変更

ras0q committed:
Merge branch 'feat/traqbot' of github.com:hackathon-21-spring-02/back-end into feat/traqbot

ras0q committed:
♻️: 関数名変更

ras0q committed:
🚀: deleteを一括で

ras0q committed:
♻️: 変数名変更

ras0q committed:
🚀: insertを一括で

SSlime-s commented on model/file.go:
```diff
@@ -196,6 +208,57 @@ func ToggleFileFavorite(ctx context.Context, accessToken string, userID string,
+func InsertFiles(ctx context.Context, files []*File) error {
+	_, err := db.NamedExecContext(
+		ctx,
+		"INSERT IGNORE INTO files (id, title, composer_id, composer_name, message_id, created_at) "+
+			"VALUES (:id, :title, :composer_id, :composer_name, :message_id, :created_at)",
```
ここ分けてる意味あるの？

ras0q replied:
長かったからしか理由ないけどバッククォートで書いた方が良さそう
一行はさすがに見にくくない？
https://stackoverflow.com/questions/36244767/long-queries-in-golang

SSlime-s replied:
はい `` ` ``使ってもろて
複数行に分けるために `+` 使うの違くないか という主張でもあった

SSlime-s commented on router/bot/message_created.go:
```diff
@@ -0,0 +1,46 @@
+				CreatedAt:    payload.Message.CreatedAt,
+			})
+
+			err = model.InsertFiles(ctx, insertReq)
```
ループの中で叩いたらだめでは

SSlime-s commented on router/bot/message_updated.go:
```diff
@@ -0,0 +1,65 @@
+				CreatedAt:    payload.Message.CreatedAt,
+			})
+
+			err = model.InsertFiles(ctx, insertReq)
```
ここも

SSlime-s reviewed (changes requested):
ちょっと

ras0q committed:
🐛  insertFilesをfor文の中で叩いていた

ras0q committed:
🐛  いらない改行

ras0q committed:
🎨  queryの書き方を変えた

Ras96 commented:
なおした

ras0q committed:
♻️  クエリを直接引数に

ras0q committed:
🗑️  コメント消した

SSlime-s commented on model/file.go:
```diff
@@ -196,6 +208,60 @@ func ToggleFileFavorite(ctx context.Context, accessToken string, userID string,
+}
+
+func InsertFiles(ctx context.Context, files []*File) error {
+	query := `
+		INSERT IGNORE INTO files (id, title, composer_id, composer_name, message_id, created_at)
+		VALUES (:id, :title, :composer_id, :composer_name, :message_id, :created_at)`
+
```
`` ` `` だけの行から始まるの :yaya: だけど見やすいからいいといえばいい
(好み)

SSlime-s commented on model/file.go:
```diff
@@ -196,6 +208,60 @@ func ToggleFileFavorite(ctx context.Context, accessToken string, userID string,
+
+	_, err := db.NamedExecContext(
+		ctx,
+		query,
```
ここ別に直接書いてもいいのではと思ってしまった

SSlime-s commented on router/bot/message_updated.go:
```diff
@@ -0,0 +1,64 @@
+		return err
+	}
+
+	deleteReq := make([]string, 0, len(oldArr)) // 削除するメッセージの配列
```
メッセージの配列ではなくないか
あと自明なコメントはなくても良い気がする + Req って変数名ちょっと :yaya: じゃない？
deleteFileIds (これは長いけど) とかのが良い気がせん？

SSlime-s reviewed (commented):
ちょっと

Ras96 commented:
@SSlime-s 
なおした(時計ずれててコミットとコメントの順番おかしいことになってるけど)

SSlime-s commented on router/bot/message_updated.go:
```diff
@@ -50,13 +50,13 @@ func MessageUpdatedHandler(ctx context.Context, accessToken string, payload *tra
 	}
 
-	deleteReq := make([]string, 0, len(oldArr)) // 削除するメッセージの配列
+	deleteFields := make([]string, 0, len(oldArr))
```
ここでの Fields ってどういう意味？

ras0q replied:
deleteFileIdsと見間違えてた:weary:

ras0q committed:
🐛  typo

ras0q committed:
Merge branch 'main' of ssh://github.com/hackathon-21-spring-02/back-end into feat/traqbot

Ras96 commented:
なおした

SSlime-s reviewed (approved):
過去の自分を信じて
